flask_jwt_extended/internal_utils.py

```diff
@@ -51,3 +64,31 @@ def custom_verification_for_token(jwt_header: dict, jwt_data: dict) -> None:
     if not jwt_manager._token_verification_callback(jwt_header, jwt_data):
         error_msg = "User claims verification failed"
         raise UserClaimsVerificationError(error_msg, jwt_header, jwt_data)
+
+
+def get_json_encoder(app: Flask) -> Type[json.JSONEncoder]:
+    """Get the JSON Encoder for the provided flask app
+
+    Starting with flask version 2.2 the flask application provides a
+    interface to register a custom JSON Encoder/Decoder under the json_provider_class.
+    As this interface is not compatible with the standard JSONEncoder, the `default`
+    method of the class is wrapped.
+
+    Lookup Order:
+      - app.json_encoder - For Flask < 2.2
+      - app.json_provider_class.default
+      - flask.json.provider.DefaultJSONProvider.default
+
+    """
+    if not HAS_JSON_PROVIDER:  # pragma: no cover
+        return app.json_encoder
+
+    # If the registered JSON provider does not implement a default classmethod
+    # use the method defined by the DefaultJSONProvider
+    default = getattr(app.json_provider_class, "default", DefaultJSONProvider.default)
+
+    class JSONEncoder(json.JSONEncoder):
+        def default(self, o: Any) -> Any:
+            return default(o)  # pragma: no cover
```

Comment:
My only worry here is that I'm not sure if these ad-hoc created classes are going to get properly cleaned up in memory since one is basically being created every time we call `config.json_encoder`. I would hate to introduce a memory leak that caused problems in prod. Even though I agree that this looks prettier, I think maybe moving this back to the `try` statement at the top of the file would be safer. 

Thoughts?

Reply:
As we want to respect the `app.json_provider_class` I don't see a way to do this at import time as `current_app` is not valid then. I'm still thinking about a clean way, I'm not 100% happy with this either.

Reply:
Something like this:
````python
class JSONEncoder(json.JSONEncoder):
    def default(self, o: Any) -> Any:
        # If the registered JSON provider does not implement a default classmethod
        # use the method defined by the DefaultJSONProvider
        default = getattr(current_app.json_provider_class, "default", DefaultJSONProvider.default)
        return default(o)


def get_json_encoder(app: Flask) -> Type[json.JSONEncoder]:
    if not HAS_JSON_PROVIDER:  # pragma: no cover
        return app.json_encoder

    return JSONEncoder
````

Reply:
If it works, yeah!

Reply:
Seems to work. Added two tests (one for flask < 2.2, one for flask >= 2.2) to ensure this is working.